generic_expression_patterns_modules/lv.py

```diff
@@ -0,0 +1,546 @@
+        (
+            summary_data[summary_data["Rank (simulated)"] < generic_threshold]
+            .set_index("Gene ID")
+            .index
+        )
+    )
+    print(f"No. of other genes: {len(ls_other_genes)}")
+
+    # Create dictionary
+    dict_genes = {
+        "generic": ls_generic_genes,
+        "other": ls_other_genes,
+    }
+
+    return dict_genes
+
+
+def process_generic_specific_gene_lists(dict_genes, LV_matrix):
+    """
+    This function returns the dictionary of generic genes and specific genes
+    that were included in the multiPLIER or eADAGE analyses. We want to make
+    sure that our gene lists obtained from SOPHIE vs multiPLIER or eADAGE
+    are consistent. This will prevent indexing by a gene that doesn't
+    exist and resulting in NA values downstream.
+
+    Arguments
+    ---------
+    dict_genes: dict
+        Dictionary mapping gene ids to label="generic", "other"
+
+    LV_matrix: df
+        Dataframe containing contribution of gene to LV (gene x LV matrix)
+    """
+    model_genes = list(LV_matrix.index)
+
+    processed_dict_genes = {}
+    for gene_label, ls_genes in dict_genes.items():
+        ls_genes_processed = list(set(model_genes).intersection(ls_genes))
+
+        processed_dict_genes[gene_label] = ls_genes_processed
+
+    return processed_dict_genes
+
+
+def get_nonzero_LV_coverage(dict_genes, LV_matrix):
+    """
+    This function counts the number of LVs that each
+    gene is present in (i.e. has a nonzero contribution).
+    This function returns a dictionary [gene id]: number of LVs
+
+    Arguments
+    ---------
+    dict_genes: dict
+        Dictionary mapping gene ids to label="generic", "other"
+
+    LV_matrix: df
+        Dataframe containing contribution of gene to LV (gene x LV matrix)
+    """
+    dict_nonzero_coverage = {}
+    for gene_label, ls_genes in dict_genes.items():
+        LV_series = (LV_matrix.loc[ls_genes] != 0).sum(axis=1)
+
+        dict_nonzero_coverage[gene_label] = LV_series
+
+    return dict_nonzero_coverage
+
+
+def get_highweight_LV_coverage(dict_genes, LV_matrix, quantile=0.9):
+    """
+    This function count the number of LVs that each
+    gene contributes a lot to (i.e. has a high negative or positive
+    weight contribution).
+    This function returns a dictionary [gene id]: number of LVs
+
+    Note: Using the quantile means that each LV has the same number
+    of high weight values. Also here we are using a quantile cutoff
+    since our distribution is not normal (exponential PDF)
+
+    Arguments
+    ---------
+    dict_genes: dict
+        Dictionary mapping gene ids to label="generic", "other"
+
+    LV_matrix: df
+        Dataframe containing contribution of gene to LV (gene x LV matrix)
+
+    quantile: float(0,1)
+        Quantile to use to threshold weights. Default set to 90th quantile.
+    """
+    thresholds_per_LV = LV_matrix.quantile(quantile)
+
+    # Manually checked that genes selected as high weight
```

Comment:
This repo mainly has just one integrated test to test that all my scripts are able to run, but I definitely want to start adding more unit tests throughout my pipelines especially as they get larger. Since this project is wrapping up I'm ok having these tests as comments for my future self. But for future I would like to start making similar tests to what you have implemented.